### DISCUSSION TITLE
Double click to Import

### DISCUSSION THREAD

**Discussion by jlacroix10** [ppy/osu#22586] [Ideas] [answered]
So, when importing files via the import files in the debug section of the settings menu, you need to select the file you want to import, and then click the import button. This can be quite cumbersome, especially when you need to import multiple files, like when you are importing beatmaps from a beatmap pack. 

I would like to be able to double click to import a file. This is what the code to that would look like: 
https://github.com/jlacroix10/osu/compare/master...jlacroix10:osu:DoubleClickToImport?expand=1

**Comment by bdach** [comment]
I'm not opposed to adding a double-click behaviour myself but feel that doing only that and stopping there is going to be pretty poor because there is no way to know that it will do something. It's not a UX behaviour that is used anywhere else in the game. The screen has a very "thrown together" feel to it right now.

My thoughts:

![image](https://user-images.githubusercontent.com/20418176/218195484-51997239-be0c-4ee7-a839-c25bb7ea4c03.png)

With those changes applied, I feel like this proposal would be making much more sense.

**Reply by jlacroix10** [reply]
That makes a lot of sense. Did you have anything in mind for the button? 
Should I just leave it there like so?
![osu_2023-02-10_17-20-04](https://user-images.githubusercontent.com/37085907/218213778-df27d7a1-7c39-40b8-bbea-61a61a7142c3.jpg)
Or maybe it would be better to center it... In which case, how big do we want it? I don't think we want the entire width, though it works I guess. 
![osu_2023-02-10_17-33-04](https://user-images.githubusercontent.com/37085907/218213830-5f474e98-a1cb-4c88-bc38-f61b76275c23.jpg)

As for highlighting, I will have to get back to you on that. Do you think it is an option that we would want to implement at the framework level though? 

Edit: Actually, did you want to use the space from the select text (where is says the name of the file you are importing) to explain the process of importing a file?

**Reply by jlacroix10** [reply]
Rough implementation of highlighting selected file: https://github.com/ppy/osu/compare/master...jlacroix10:osu:HighlightSelectedFile?expand=1


Watch my change in action: 

https://user-images.githubusercontent.com/37085907/218225081-bfe55003-ce0f-4593-b9e0-ecf3ee52e281.mp4

I just guessed what colour to use, let me know which one you would like me to use and I'll update my branch to match.

**Reply by bdach** [reply]
![image](https://user-images.githubusercontent.com/20418176/218252357-801dcb61-26f4-453c-a5df-db369c40dd53.png)

is what I'd say.

Before we go too far down here I'd probably want to double check this with @ppy/team-client though.

**Reply by jlacroix10** [reply]
Makes sense. So at this point, should I just wait before making more changes?

**Reply by bdach** [reply]
Yes please.

**Comment by Susko3** [comment]
As you've said, this is under the debug section, so it's not really that important. And where it matters (in the Editor song/background image selector) it's one click to select and use the file.

For importing beatmaps you're not meant to use this screen, but instead drag and drop the files into osu! (that way you can drag multiple files) or use the (not yet implemented) file associations.

**Reply by jlacroix10** [reply]
I disagree with the importance of that screen. Dragging and dropping files require you to exit exclusive fullscreen mode to work (even with multiple monitors) and file associations are not yet implemented. 

Also, regardless of how important it is, if it is meant to be there, it should be well made. The code to make it look good is actually pretty straight forward, as someone with absolutely no idea of how the osu codebase works (aka me), was able to get pretty far in. Therefore, I think it is worth at least exploring how it could be improved.

**Reply by Susko3** [reply]
Exiting out of fullscreen is not necessary to drag and drop files into a window.

For example, you can Alt-Tab into osu! while holding some files and then drop them. Or, hover over the osu! taskbar button with the files, and when the game gains focus, drop them.

**Reply by jlacroix10** [reply]
Right, I guess that makes sense. I should have thought of that, thank you.

**Comment by peppy** [comment, marked as answer]
Unless this screen is to be used outside of debug, there's no need to focus on this.

If this discussion is to continue, the focus should be in the same modular dialog's other usages, such as selecting a file in the editor for use as an audio or background resource.

Behaviour should match windows or macOS. There's no need to reinvent the wheel.

**Reply by jlacroix10** [reply]
This screen is not used outside of debug. 

Selecting files in the editor works differently than in the debug section, so the same screen isn't used in both places. 

That screen uses the file selector component from the framework under the hood, so the behaviour should match windows and macOS. 

That being said, like I said in another thread, I believe that if this screen is meant to be there, it should be well made. Given that it is fairly straight forward to implement these changes (I say this because I have no experience working with the code base and I still was able to get pretty far on that screen), I believe it would be worth at least exploring how that screen could be improved.
